net/sched/sch_gkprioq.c

```diff
@@ -0,0 +1,316 @@
+ * packets enables us to give low priority to clients who quickly send many 
+ * requests, so  that we can service legitimate clients (who typically send 
+ * requests more slowly). The DoS scenario is when assigning low priorities 
+ * helps block out malicious senders, who could otherwise overwhelm Gatekeeper 
+ * with a very high number of requests, denying service to the requests of
+ * legitimate users.
+ *
+ *
+ * To assign priority to packet, Gatekeeper uses the six bits of the DSCP field
+ * as it is a sufficient and convenient way to limit the maximum number of 
+ * priorities we'd like to deal with. Additionally, using the DSCP field also 
+ * allows us to more easily use tools like iptables and tc, which already 
+ * understand fields in the IP header. 
+ *
+ *
+ * Although the DSCP field is usually used to specify DiffServe traffic 
+ * classifcation options, it is NOT used for that intent here. 
+ *
+ *
+ *
+ *   The Gatekeper standalone codebase is found here: 
+ *	          
+ *          https://github.com/AltraMayor/gatekeeper
+ *
+ *
+ */
+
+
+struct gkprioq_sched_data {
+/* Parameters */
+	u32 max_limit;
+
+/* Variables */
+	struct Qdisc *prio_array[MAX_PRIORITY];
+	u16 highest_prio;
+};
+
+static inline u16 calc_new_high_prio(const struct gkprioq_sched_data *q)
+{
+	int prio;
+
+	for (prio = q->highest_prio - 1; prio > 0; prio--) {
+		if ((q->prio_array[prio])->q.qlen > 0)
+			return prio;
+	}
+
+	return 0; /* Queue 0 could be non empty OR GK queue is empty. In any
+		   * case return 0 (default highest priority setting). */
+}
+
+static int gkprioq_enqueue(struct sk_buff *skb, struct Qdisc *sch,
+			   struct sk_buff **to_free)
+{
+	struct Qdisc *your_queue;
+	int prio, ret, wlen, gk_len;
+	u16 your_prio; 
+
+	struct gkprioq_sched_data *q = qdisc_priv(sch);
+	wlen = skb_network_offset(skb);
+
+	switch(tc_skb_protocol(skb)) {
+	case htons(ETH_P_IP):
+		wlen += sizeof(struct iphdr);
+		if (!pskb_may_pull(skb, wlen))
+			goto drop;
+
+		your_prio = (ipv4_get_dsfield(ip_hdr(skb))) >> 2;
+		break;
+
+	case htons(ETH_P_IPV6):
+	 	wlen += sizeof(struct ipv6hdr);
+		if (!pskb_may_pull(skb, wlen))
+			goto drop;
+
+		your_prio = (ipv6_get_dsfield(ipv6_hdr(skb))) >> 2;
+	      	break;
+
+	default: 
+		your_prio = 0;
+		break;
+	}
+
+	if (your_prio > q->highest_prio)
+		q->highest_prio = your_prio; /* Checking if a new highest 
+					      * priority. */
+			
+	your_queue = q->prio_array[your_prio];
+
+	gk_len = sch->q.qlen;
+	if (gk_len < q->max_limit) {
+
+		ret = qdisc_enqueue(skb, your_queue, to_free);
+		if (ret != NET_XMIT_SUCCESS) {
+			if (net_xmit_drop_count(ret))
+				qdisc_qstats_drop(sch);
+		return ret;
+		}
+
+		qdisc_qstats_backlog_inc(sch, skb);
+		sch->q.qlen++;
+		return NET_XMIT_SUCCESS;
+	}
+
+	/* your_queue is taking up all of GK, no choice but to remove a packet
+	 * and then enqueue the ingressed packet.
+	 */
+
+	else if (your_queue->q.qlen == your_queue->limit || your_prio == 0) {
```

Comment:
Hi Michel,  yes I was not confident with `else if` because I didn't know if I should drop the ingress packet or a packet already enqueued, before enqueueing another packet. Will make the change.